src/core/lispStream.cc

```diff
@@ -566,16 +566,76 @@ int stream_set_column(T_sp stream, int column) {
 CL_LISPIFY_NAME("gray:%stream-start-line-p")
 CL_DEFUN bool stream_start_line_p(T_sp stream) {
   return stream.isA<AnsiStream_O>() ? stream.as_unsafe<AnsiStream_O>()->start_line_p()
-                                    : T_sp(eval::funcall(gray::_sym_stream_line_column, stream)).notnilp();
+                                    : T_sp(eval::funcall(gray::_sym_stream_start_line_p, stream)).notnilp();
+}
+
+// This function is exposed to CL because it is needed to implement
+// the GRAY:STREAM-LINE-NUMBER for ansi-stream. The stream argument is
+// guaranteed to be an AnsiStream_sp so recursion is avoided.
+CL_LISPIFY_NAME("gray:%stream-line-number")
+CL_DEFUN T_sp stream_output_line(T_sp stream) {
+  return stream.isA<AnsiStream_O>() ? stream.as_unsafe<AnsiStream_O>()->output_line()
+                                    : eval::funcall(gray::_sym_stream_line_number, stream);
+}
+
+uint stream_output_column_as_uint(T_sp stream) {
+  T_sp column = stream_output_column(stream);
+  return gc::IsA<Real_sp>(column) ? clasp_to_integral<uint>(clasp_floor1(gc::As_unsafe<Real_sp>(column))) : 0;
+}
+
+uint stream_output_line_as_uint(T_sp stream) {
+  T_sp line = stream_output_line(stream);
+  return gc::IsA<Real_sp>(line) ? clasp_to_integral<uint>(clasp_floor1(gc::As_unsafe<Real_sp>(line))) : 1;
+}
+
+void stream_update_output_cursor(T_sp stream, claspCharacter c) {
+  if (stream.isA<AnsiStream_O>())
+    stream.as_unsafe<AnsiStream_O>()->update_output_cursor(c);
+}
+
+void stream_restore_output_cursor(T_sp stream) {
+  if (stream.isA<AnsiStream_O>())
+    stream.as_unsafe<AnsiStream_O>()->restore_output_cursor();
+}
+
+// Stream input tracking functions
+
+// This function is exposed to CL because it is needed to implement
+// the GRAY:STREAM-INPUT-COLUMN for ansi-stream. The stream argument
+// is guaranteed to be an AnsiStream_sp so recursion is avoided.
+CL_LISPIFY_NAME("gray:%stream-input-column")
```

Comment:
I'm mildly concerned about these functions being exposed in the GRAY package. I see that they are unexported later, but maybe it would be cleaner to move them to CORE, or for them to be defined in GRAY but not exported (I think there's a C macro to use instead of `SYMBOL_EXPORT_SC_` for that?)

Reply:
I moved them from core because only gray is using them and they seem to be an internal design detail of the gray package. The scrapper and expose_function export all function automatically. I started down the patch of adding a "internal" option, but it looks like that will take more work and is kind of a tangential issue.